// @zaiusinc/app-sdk/dist/app/types/AppManifest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VENDOR_FORMAT = exports.VERSION_FORMAT = exports.APP_ID_FORMAT = exports.ReachabilityMetric = exports.DisengagementMetric = exports.AttributableMetric = exports.EngagementMetric = exports.DeliveryMetric = exports.ChannelType = void 0;
var ChannelType;
(function (ChannelType) {
    ChannelType["Email"] = "email";
    ChannelType["AppPush"] = "app_push";
    ChannelType["WebPush"] = "web_push";
    ChannelType["WebModal"] = "web_modal";
    ChannelType["WebEmbed"] = "web_embed";
    ChannelType["SMS"] = "sms";
    ChannelType["Api"] = "api";
    ChannelType["DirectMail"] = "direct_mail";
    ChannelType["WhatsApp"] = "whatsapp";
    ChannelType["FacebookMessenger"] = "facebook_messenger";
    ChannelType["Ad"] = "ad";
    ChannelType["SegmentSync"] = "segment_sync";
    ChannelType["TestChannel"] = "test_channel";
})(ChannelType || (exports.ChannelType = ChannelType = {}));
var DeliveryMetric;
(function (DeliveryMetric) {
    DeliveryMetric["Sent"] = "sent";
    DeliveryMetric["Delivery"] = "delivery";
    DeliveryMetric["DeliveryUnknown"] = "delivery_unknown";
})(DeliveryMetric || (exports.DeliveryMetric = DeliveryMetric = {}));
var EngagementMetric;
(function (EngagementMetric) {
    EngagementMetric["Open"] = "open";
    EngagementMetric["Click"] = "click";
    EngagementMetric["Engage"] = "engage";
})(EngagementMetric || (exports.EngagementMetric = EngagementMetric = {}));
var AttributableMetric;
(function (AttributableMetric) {
    AttributableMetric["Open"] = "open";
    AttributableMetric["Click"] = "click";
    AttributableMetric["Engage"] = "engage";
    AttributableMetric["Delivery"] = "delivery";
})(AttributableMetric || (exports.AttributableMetric = AttributableMetric = {}));
var DisengagementMetric;
(function (DisengagementMetric) {
    DisengagementMetric["Disengage"] = "disengage";
    DisengagementMetric["OptOut"] = "opt-out";
    DisengagementMetric["SpamReport"] = "spam_report";
    DisengagementMetric["ListUnsubscribe"] = "list_unsubscribe";
})(DisengagementMetric || (exports.DisengagementMetric = DisengagementMetric = {}));
var ReachabilityMetric;
(function (ReachabilityMetric) {
    ReachabilityMetric["HardBounce"] = "hard_bounce";
    ReachabilityMetric["SoftBounce"] = "soft_bounce";
})(ReachabilityMetric || (exports.ReachabilityMetric = ReachabilityMetric = {}));
exports.APP_ID_FORMAT = /^[a-z][a-z_0-9]{2,31}$/;
exports.VERSION_FORMAT = /^\d+\.\d+\.\d+(-(((dev|beta)(\.\d+)?)|private))?$/;
exports.VENDOR_FORMAT = /^[a-z0-9]+(_[a-z0-9]+)*$/;
