// @zaiusinc/app-sdk/dist/app/lib/AuthorizationGrantResult.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthorizationGrantResult = void 0;
const FormResult_1 = require("./FormResult");
/**
 * Used to compose a response to the onAuthorizationGrant lifecycle request
 */
class AuthorizationGrantResult extends FormResult_1.FormResult {
    redirectSection;
    /**
     * @param redirectSection the section of the settings form the user will be redirected to
     */
    constructor(redirectSection) {
        super();
        this.redirectSection = redirectSection;
    }
    /**
     * Add an error to display to the user for a particular form field (implicitly scoped to the redirected section)
     * @param field key to display the error under, as defined in the form schema
     * @param error message to display to the user
     */
    addError(field, error) {
        return this.addErrorInternal(`${this.redirectSection}.${field}`, error);
    }
    /**
     * @hidden
     * Used internally to get the complete response
     */
    getResponse() {
        return {
            errors: this.errors,
            toasts: this.toasts,
            redirect: this.redirectSection,
            redirectMode: 'settings'
        };
    }
}
exports.AuthorizationGrantResult = AuthorizationGrantResult;
