// @zaiusinc/app-sdk/dist/app/lib/ChannelTargetResult.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelTargetResult = void 0;
const FormResult_1 = require("./FormResult");
/**
 * Result of {@link Channel.target}.
 */
class ChannelTargetResult extends FormResult_1.FormResult {
    targeting;
    /**
     * @param targeting the complete set of targeting requirements, if already known
     */
    constructor(targeting) {
        super();
        this.targeting = targeting || [];
    }
    /**
     * Add a single targeting requirement
     * @param targeting requirement to add
     */
    addTargeting(targeting) {
        this.targeting.push(targeting);
        return this;
    }
    /**
     * Provide the complete set of targeting requirements
     * @param targeting requirements to provide
     */
    setTargeting(targeting) {
        this.targeting = targeting;
        return this;
    }
    /**
     * Add an error to display to the user for a particular form field (implicitly scoped to the content settings form)
     * @param section the section within the form that the error applies to
     * @param field the field within the section that the error applies to
     * @param error message to display to the user
     */
    addError(section, field, error) {
        return this.addErrorInternal(`${section}.${field}`, error);
    }
    /**
     * @hidden
     * Used internally to get the complete response
     */
    getResponse() {
        if ((!this.targeting || this.targeting.length === 0) &&
            Object.keys(this.errors).length === 0 &&
            this.toasts.length === 0) {
            this.addToast('danger', 'Failed to determine targeting requirements');
        }
        return {
            targeting: this.targeting,
            errors: this.errors,
            toasts: this.toasts
        };
    }
}
exports.ChannelTargetResult = ChannelTargetResult;
