// @zaiusinc/app-sdk/dist/app/lib/Headers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Headers = void 0;
class Headers {
    headers = {};
    constructor(headers = []) {
        this.addFromArray(headers);
    }
    get(name) {
        return this.headers[name.toLowerCase()];
    }
    has(name) {
        return !!this.headers[name.toLowerCase()];
    }
    add(name, value) {
        if (typeof value !== 'string' || value === '') {
            return;
        }
        name = name.toLowerCase();
        if (this.headers[name]) {
            this.headers[name] += ',' + value;
        }
        else {
            this.headers[name] = value;
        }
    }
    set(name, value) {
        name = name.toLowerCase();
        if (value === null || value === '' || typeof value !== 'string') {
            delete this.headers[name];
        }
        else {
            this.headers[name] = value;
        }
    }
    toArray() {
        return Object.keys(this.headers).map((key) => [key, this.headers[key]]);
    }
    addFromArray(headers) {
        headers.forEach(([name, value]) => this.add(name, value));
    }
    setFromObject(headers) {
        Object.keys(headers).forEach((name) => this.set(name, headers[name]));
    }
}
exports.Headers = Headers;
