// @zaiusinc/app-sdk/dist/app/lib/LiquidExtensionResult.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiquidExtensionResult = void 0;
/**
 * Result of {@link LiquidExtension.perform}. Clients should use the static {@link success} and {@link error} helper
 * methods (as opposed to constructing manually).
 */
class LiquidExtensionResult {
    success;
    output;
    message;
    /**
     * Builds a success response.
     * @param output the resulting output (must be JSON-serializable)
     */
    static success(output) {
        return new LiquidExtensionResult(true, output);
    }
    /**
     * Builds an error response.
     * @param message
     */
    static error(message) {
        return new LiquidExtensionResult(false, undefined, message);
    }
    /**
     * @hidden
     * Use the static helper methods instead.
     */
    constructor(success, output, message) {
        this.success = success;
        this.output = output;
        this.message = message;
    }
    /**
     * @hidden
     * Used internally to get the complete response
     */
    getResponse() {
        return {
            output: this.success ? JSON.stringify(this.output === undefined ? null : this.output) : undefined,
            error: this.success ? undefined : this.message || 'unknown'
        };
    }
}
exports.LiquidExtensionResult = LiquidExtensionResult;
