// @zaiusinc/app-sdk/dist/app/lib/Request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Request = void 0;
const Headers_1 = require("./Headers");
class Request {
    method;
    path;
    params;
    headers;
    bodyData;
    parsedJsonBody;
    constructor(method, path, params, headers, body) {
        this.method = method;
        this.path = path;
        this.params = params;
        this.headers = new Headers_1.Headers(headers);
        this.bodyData = body || new Uint8Array(0);
    }
    get body() {
        return this.bodyData;
    }
    get bodyJSON() {
        if (this.parsedJsonBody) {
            return this.parsedJsonBody;
        }
        if (this.bodyData.length === 0) {
            return null;
        }
        // TODO: attempt to read character set from content-type header and/or allow user to set the encoding
        return (this.parsedJsonBody = JSON.parse(Buffer.from(this.bodyData).toString('utf8')));
    }
    get contentType() {
        const value = this.headers.get('content-type');
        if (value) {
            return value.split(';')[0];
        }
        return null;
    }
}
exports.Request = Request;
