// @zaiusinc/app-sdk/dist/app/lib/Response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Response = void 0;
const Headers_1 = require("./Headers");
class Response {
    status = 0;
    headers = new Headers_1.Headers();
    bodyData;
    constructor(status, bodyJSON, headers) {
        if (status !== undefined) {
            this.status = status;
        }
        if (bodyJSON !== undefined) {
            this.bodyJSON = bodyJSON;
        }
        if (headers !== undefined) {
            this.headers = headers;
        }
    }
    set body(body) {
        if (body === null || typeof body !== 'string') {
            this.bodyData = undefined;
        }
        else {
            this.bodyData = new Uint8Array(Buffer.from(body));
        }
    }
    get bodyAsU8Array() {
        return this.bodyData;
    }
    set bodyAsU8Array(body) {
        this.bodyData = body;
    }
    set bodyJSON(body) {
        if (!this.headers.has('content-type')) {
            this.headers.set('content-type', 'application/json');
        }
        if (body !== undefined) {
            this.bodyData = new Uint8Array(Buffer.from(JSON.stringify(body)));
        }
        else {
            this.bodyData = undefined;
        }
    }
}
exports.Response = Response;
